### PR TITLE
zkVM: fix build issues arising from undefined `_end`

### PR DESCRIPTION
**Pull request by SchmErik**
When building zkvm as a host-side staticlib to link with platforms such as go, the build can fail due to an undefined `_end` symbol. Hide this when building for the host.

Fixes: #1329

### CODE REVIEW

File: risc0/zkvm/src/lib.rs [hunk 1]
```diff
@@ -71,6 +71,7 @@
 
 extern crate alloc;
 
+#[cfg(target_os = "zkvm")]
```
Comment (flaub) [inline risc0/zkvm/src/lib.rs:74]: I think this breaks the ability to use rust-analyzer

Comment (SchmErik) [reply risc0/zkvm/src/lib.rs]: Thinking about this some more, the original bug report and issues we've seen internally point to the undefined `_end` symbol used in `sys_alloc_aligned`. I'm thinking that it's probably better to revert all of these changes and put cfg's around where _end appears... Let me know if you have thoughts on this. I'm going to make this change and change the title/description of this PR.

**Comment by Wollac** [comment]
It seems strange to me that the `_end` is only hidden on non-zkvm targets. This means that `sys_alloc_aligned` is still compiled and exported, but is probably incorrect because parts of it are changed.

I actually preferred the previous approach where no syscalls were exported or even compiled on the host as they are not needed here. Can we maybe keep this and just make sure that the `guest` mod is not behind a target flag? 
If not, I think it would still be better to disable the entire `sys_alloc_aligned` function on non-zkvm targets (but nothing else). Rather than have it compile a somehow modified version.

**Comment by flaub** [comment]
It's nice to have the IDE do analysis and references. Thus the guest code is still meant to be able to be compiled on a target that it won't run on.

**Review by flaub — APPROVED** [review]

**Pull request merged**